### PR TITLE
Support Dev Containers / Github Codespaces

### LINKED ISSUE #113019: Add Codespaces configuration dev container
### What would you like to be added?

For contributors to Kubernetes, a configuration to assure that Codespaces will work with Kubernetes will enable the use of a cloud-based development environment to easily provision an environment with all the requirements and dependencies automatically installed and configured. This will make life easy for new and existing contributors when they are working on multiple branches with different dependencies, or don't have access to a computer or network powerful enough to build Kubernetes.

The Codespaces configuration consists of a Development Container ([devcontainer](https://containers.dev)) file `devcontainer.json` which encodes the configuration of the environment.

### Why is this needed?

sig/contribex is working on a new self-paced course for Kubernetes contribution to be hosted on k8s.dev. Codespaces will be one option for working through the course material

### PR DESCRIPTION
**Pull request by Sharpz7**
Based on: https://github.com/craiglpeters/kubernetes-devcontainer

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

This PR adds support for devcontainers. Right now, it will take a significant amount of time to build - but ideally I would want to also create an image that is built as part of this, maybe in its own repo, that could then be pulled to speed up build times.
Also a wider discussion on whether you would want to support devcontainers all over k8s should probably happen.

https://github.com/kubernetes/test-infra/issues/30454
https://github.com/kubernetes/test-infra/pull/30511#discussion_r1305712447

~~Right now, I have added a custom golang install feature that can take advantage of `.go-version` being present, but maybe there is a more elegant solution we can use?~~
I tried this and it didn't work - my bet is the devcontainer is being created before the repo is present so the version cannot be found. This means we now have the version stated in two places... not ideal.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/113019

#### Does this PR introduce a user-facing change?

```release-note
Adds the code/configs to allow for users to use dev-containers when developing k8s. This brings support for GitHub Codespaces, EnvBuilder, the devcontainer-cli and more.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [Usage]: https://github.com/features/codespaces
- [More Info]: https://containers.dev/
```

/kind feature
/assign @Sharpz7
/sig architecture
/sig contributor-experience

cc @craiglpeters

### CODE REVIEW

**Comment by mrbobbytables** [comment]
This PR replaces #120551 and addresses the feedback I left in https://github.com/kubernetes/kubernetes/pull/120551#pullrequestreview-1688307879 (adding an owners file)

I verified that it works and I was successfully able to build kubernetes.

/assign dims cblecker
/priority important-soon
/triage accepted

**Comment by mrbobbytables** [comment]
and I forgot to lgtm in my last comment 🤦 

/lgtm

File: .devcontainers/OWNERS
Comment (dims) [inline .devcontainers/OWNERS:5]: Why is this necessary?

Comment (mrbobbytables) [reply .devcontainers/OWNERS:5]: That was my suggestion to mirror the .github directory that contribex maintains. We could potentially remove it as the root approvers are also well positioned to look at and test the changes

Comment (dims) [reply .devcontainers/OWNERS:5]: yes pleqase

Comment (cblecker) [reply .devcontainers/OWNERS:5]: +1

File: .devcontainers/devcontainer.json
Comment (cblecker) [inline .devcontainers/devcontainer.json:5]: Is this something we'll need to bump in the future? If so, under what circumstances?

File: .devcontainers/devcontainer.json
Comment (cblecker) [inline .devcontainers/devcontainer.json:21]: Same here, does this need to be consistent with the go version we are using in the rest of the repo?

File: .devcontainers/devcontainer.json
Comment (cblecker) [inline .devcontainers/devcontainer.json:24]: If we are developing against the code base, should we be building kubectl from source?

File: .devcontainers/devcontainer.json
Comment (cblecker) [inline .devcontainers/devcontainer.json:11]: Is this run on the local system or inside the container?

File: .devcontainers/devcontainer.json
Comment (cblecker) [reply .devcontainers/devcontainer.json:5]: Should we be using kube-cross here? (`registry.k8s.io/build-image/kube-cross`)

**Review by cblecker — CHANGES_REQUESTED** [review]
Some questions before we move forward
/hold

File: .devcontainers/OWNERS
Comment (cblecker) [inline .devcontainers/OWNERS:1]: Also, this entire directory should be `.devcontainer` not `.devcontainers`

**Review by cblecker — CHANGES_REQUESTED** [review]

File: .devcontainers/devcontainer.json
Comment (cblecker) [reply .devcontainers/devcontainer.json:5]: Looking at https://containers.dev/guide/dockerfile, it looks like we could possibly reference the `../build/build-image/Dockerfile`?

Comment (cblecker) [reply .devcontainers/devcontainer.json:5]: maybe something like:
```
    "build": {
        "dockerfile": "../build/build-image/Dockerfile",
        "args": {
            "KUBE_CROSS_IMAGE": "registry.k8s.io/build-image/kube-cross",
            "KUBE_CROSS_VERSION": "v1.29.0-go1.21.3-bullseye.0"
        }
    },
```

File: .devcontainers/devcontainer.json
Comment (cblecker) [reply .devcontainers/devcontainer.json:21]: If we use our own base image, is this required?

File: .devcontainers/OWNERS
Comment (Sharpz7) [reply .devcontainers/OWNERS:1]: I think either is fine. Is there a particular reason you would want this or is it a nit?

Just because in the future we could have more than one.

File: .devcontainers/devcontainer.json
Comment (Sharpz7) [reply .devcontainers/devcontainer.json:11]: Inside the container.

**Comment by Sharpz7** [comment]
@cblecker just a general catch-all for your points.

They are all valid, devcontainers is such a wide standard that anything is possible. But, I would argue that things like our own Docker Image, and building kubectl from source, slow down the build significantly. There is a trade-off here between speed, closeness to ground-truth, and number of files (and how many locations those files are in).

I think the idea is this would be the initial run, and then we would improve what we do overtime. For now, I think this is mostly okay - you can follow the guides easily.

File: .devcontainers/devcontainer.json
Comment (Sharpz7) [reply .devcontainers/devcontainer.json:5]: I think this is a good idea, and something we could use.

File: .devcontainers/devcontainer.json
Comment (Sharpz7) [reply .devcontainers/devcontainer.json:21]: We struggled with this. The problem being that this gets installed before the repo is downloaded. I could make a custom feature that pre-downloads the repo to check the right go version, but I thought that was a lot for an initial commit?

File: .devcontainers/devcontainer.json
Comment (Sharpz7) [reply .devcontainers/devcontainer.json:24]: I think that would slow down the build a lot? We want to try and get that balance right.

Happy to change that if it is a priority

File: .devcontainers/devcontainer.json
Comment (craiglpeters) [reply .devcontainers/devcontainer.json:21]: Because `kube-cross` includes the correct go version, the go feature would not be needed. A win!

File: .devcontainers/OWNERS
Comment (craiglpeters) [reply .devcontainers/OWNERS:1]: I think `.devcontainer` is the appropriate name as that is consistent with the directory containing the `devcontainer.json` spec.

Comment (cblecker) [reply .devcontainers/OWNERS:1]: vscode doesn't pick up the directory if it's plural

**Comment by cblecker** [comment]
@Sharpz7 The tradeoff that I'm considering here is sustainability. I would rather us have something that is slow, but that keeps updated over time as opposed to something that is fast but needs regular maintenance to keep it up to date. I'm all for iterating over time to improve things, but I'd rather it be "hey this always works, but is slow.. maybe we can make it faster" rather than "hey this stopped working.. oh someone needs to go PR in a change to update an image/version/etc to make it useful again".

**Comment by Sharpz7** [comment]
Since we are optimising:

- I would like to try and not need the "local features", we should use scripts already in the repo (like with etcd for example).
- I want the go version to be found automatically, its still an env var we would have the manually bump.
- I am 50/50 on whether we build kubectl from source. I think it boils down to what would the average developer expect? I would lean towards and installed version of kubectl.

File: .devcontainer/OWNERS
Comment (cblecker) [inline .devcontainer/OWNERS:3]:
```suggestion
```

File: .devcontainer/OWNERS
Comment (cblecker) [inline .devcontainer/OWNERS:15]:
```
+++ command: bash "hack/make-rules/../../hack/verify-yamlfmt.sh"
YAML files need to be formatted
Not currently on any branch.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   .devcontainer/OWNERS
no changes added to commit (use "git add" and/or "git commit -a")
diff --git a/.devcontainer/OWNERS b/.devcontainer/OWNERS
index ed137ad2679..4ee211a01fe 100644
--- a/.devcontainer/OWNERS
+++ b/.devcontainer/OWNERS
@@ -12,4 +12,4 @@ reviewers:
 approvers:
   - sig-contributor-experience-approvers
 labels:
-  - sig/contributor-experience
\ No newline at end of file
+  - sig/contributor-experience
Please run 'hack/update-yamlfmt.sh'
!!! [1103 05:17:16] Call tree:
!!! [1103 05:17:16]  1: hack/make-rules/../../hack/verify-yamlfmt.sh:30 kube::verify::generated(...)
+++ exit code: 1
+++ error: 1
[0;31mFAILED   verify-yamlfmt.sh	5s
```

**Review by cblecker — COMMENTED** [review]
Attempted to run this and it's failing. The build image is normally wrapped with a bunch of bash to ensure the right variables are passed in.

File: .devcontainer/devcontainer.json
Comment (craiglpeters) [inline .devcontainer/devcontainer.json:5]: Rather then referencing the docerfile, and doing a duplicate build, you can simply reference the already built image from the registry
`"image": "registry.k8s.io/build-image/kube-cross:v1.29.0-go1.21.3-bullseye.0",`
You can see a working example here: https://github.com/craiglpeters/kubernetes-devcontainer/blob/kube-cross/.devcontainer/kube-cross/devcontainer.json#L7

File: .devcontainer/devcontainer.json
Comment (craiglpeters) [inline .devcontainer/devcontainer.json:25]: Because the `kube-cross` image already has all the tools needed (except for pyyaml) you can remove these features.

File: .devcontainer/local-features/etcd/devcontainer-feature.json
Comment (craiglpeters) [inline .devcontainer/local-features/etcd/devcontainer-feature.json:1]: Not needed if using `kube-cross`

File: .devcontainer/local-features/etcd/install.sh
Comment (craiglpeters) [inline .devcontainer/local-features/etcd/install.sh:1]: Not needed if using `kube-cross`

File: .devcontainer/local-features/kubectl-kind/devcontainer-feature.json
Comment (craiglpeters) [inline .devcontainer/local-features/kubectl-kind/devcontainer-feature.json:1]: Not needed if using `kube-cross`

File: .devcontainer/local-features/kubectl-kind/install.sh
Comment (craiglpeters) [inline .devcontainer/local-features/kubectl-kind/install.sh:1]: Not needed if using `kube-cross`

File: .devcontainer/local-features/kubetest2/devcontainer-feature.json
Comment (craiglpeters) [inline .devcontainer/local-features/kubetest2/devcontainer-feature.json:1]: Not needed if using `kube-cross`

File: .devcontainer/local-features/kubetest2/install.sh
Comment (craiglpeters) [inline .devcontainer/local-features/kubetest2/install.sh:1]: Not needed if using `kube-cross`

**Review by craiglpeters — COMMENTED** [review]
Using `kube-cross` is as simple as pointing to the image in the container registry. The local features are not needed as `kube-cross` already has everything except pyyaml.

**Comment by Sharpz7** [comment]
I think using https://containers.dev/implementors/json_reference/#variables-in-devcontainerjson might be the way we can configure for the correct go version and kuberentes version, but I am not sure.

Thoughts @craiglpeters ?

**Comment by bells17** [comment]
Hello,

I've created a new PR with permission from @craiglpeters . 
May I request a review for this PR?

https://github.com/kubernetes/kubernetes/pull/124584

https://kubernetes.slack.com/archives/C1TU9EB9S/p1714223550391479?thread_ts=1713701062.612009&cid=C1TU9EB9S

**Pull request closed without merging**
